cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+            )
+
+        if self.cache:
+            images = self.cache._images(indices)
+        else:
+            images = self._images(indices).astype(self.dtype)
+        return torch.tensor(images)
+
+    def _images(self, indices: np.ndarray) -> np.ndarray:
+        raise NotImplementedError("Subclasses must implement this")
+
+    def chunks(self, chunksize=1000):
+        return ImageSourceIterator(self, chunksize=chunksize)
+
+
+class ArraySource(ImageSource):
+    def __init__(self, data: np.ndarray):
+        if data.ndim == 2:
+            data = data[np.newaxis, ...]
+        nz, ny, nx = data.shape
+        assert ny == nx, "Only square arrays supported"
+        self.data = data
+
+        super().__init__(L=ny, n=nz)
+
+    def _images(self, indices: np.ndarray):
+        return self.data[indices, ...]
+
+
+class MRCFileSource(ImageSource):
+    def __init__(self, filepath: str, *args, **kwargs):
+        from cryodrgn.mrc import MRCHeader
+
+        header = MRCHeader.parse(filepath)
+        self.header = header
+        self.mrcfile_path = filepath
+        self.dtype = header.dtype
+        self.start = 1024 + header.fields["next"]  # start of image data
+        self.nz, self.ny, self.nx = (
+            header.fields["nz"],
+            header.fields["ny"],
+            header.fields["nx"],
+        )
+        assert self.ny == self.nx, "Only square images supported"
+        self.size = self.ny * self.nx
+        self.stride = self.dtype().itemsize * self.size
+
+        super().__init__(L=self.ny, n=self.nz, filenames=filepath, *args, **kwargs)
+
+    def _images(
+        self,
+        indices: np.ndarray,
+        data: Optional[np.ndarray] = None,
+        tgt_indices: Optional[np.ndarray] = None,
+    ):
+
+        with open(self.mrcfile_path) as f:
+            if data is None:
+                data = np.empty((len(indices), self.L, self.L), dtype=self.dtype)
+                assert (
+                    tgt_indices is None
+                ), "Target indices can only be specified when passing in a preallocated array"
+                tgt_indices = np.arange(len(indices))
+            else:
+                if tgt_indices is not None:
+                    assert len(tgt_indices) == len(
+                        indices
+                    ), "indices/tgt_indices length mismatch"
+                else:
+                    tgt_indices = np.arange(len(indices))
+
+            assert isinstance(tgt_indices, np.ndarray)
+
+            for (index, tgt_index) in zip(indices, tgt_indices):
+                f.seek(self.start)
+                offset = index * self.stride
+                # 'offset' in the call below is w.r.t the current position of f
+                _data = np.fromfile(
+                    f, dtype=self.dtype, count=self.size, offset=offset
+                ).reshape(self.ny, self.nx)
+                data[tgt_index, : self.ny, : self.nx] = _data
```

Comment:
Is all this complexity due to the "preallocated fourier-size buffer" trick? This all looks sketchy to me, I would only do this if there's a big perf win.

Reply:
Removed in commit 7f3b87558d77091d39877540c5a6906de0ea41cc